base/ca/src/org/dogtagpki/server/ca/rest/CertService.java

```diff
@@ -209,12 +217,30 @@ public Response revokeCert(CertId id, CertRevokeRequest request, boolean caCert)
             if (clientCert != null) {
                 clientSerialNumber = clientCert.getSerialNumber();
                 clientSubjectDN = clientCert.getSubjectDN().toString();
-                clientRecord = processor.getCertificateRecord(clientSerialNumber);
 
-                // Verify client cert is not revoked.
-                // TODO: This should be checked during authentication.
-                if (clientRecord.getStatus().equals(ICertRecord.STATUS_REVOKED)) {
-                    throw new UnauthorizedException(CMS.getLogMessage("CMSGW_UNAUTHORIZED"));
+                String clientIssuerDN = clientCert.getIssuerDN().toString();
+                X500Name x500issuerDN = new X500Name(clientIssuerDN);
+                /*
+                 * internal revocation check only to be conducted for certs
+                 * issued by this CA
+                 * For client certs issued by external CAs, TLS mutual auth
+                 * would have completed the authenticaiton/verification if
+                 * OCSP was enabled;
+                 * Furthermore, prior to the actual revocation, client cert
+                 * is mapped against the agent group database for proper
+                 * privilege regardless of the issuer.
+                 */
+                if (x500issuerDN.equals(caX500DN)) {
```

Comment:
I thoguht the equals() method provided in X509CertImpl does exactly that, with Object level comparison.  Would "=" for Object comparison in Java be sufficient?

Reply:
The `X500Name.equals()` does compare the internal data of `x500issuerDN` and `caX500DN`, but these objects were constructed from `String`, not from the original cert object, so if they were originally encoded in a specific way, that information might be lost during conversions.

We shouldn't use `==` here since they are clearly different objects. We use `==` if we want to see if they're the same object.

I don't know enough about DN encoding, so I'll let you decide.